python-checks/src/main/java/org/sonar/python/checks/PandasModifyInPlaceCheck.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.Optional;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6734")
+public class PandasModifyInPlaceCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Do not use 'inplace=True' when modifying a dataframe.";
+
+  private static final Set<String> FULLY_QUALIFIED_EXPRESSIONS = Set.of(
+    "pandas.core.frame.DataFrame.drop",
+    "pandas.core.frame.DataFrame.dropna",
+    "pandas.core.frame.DataFrame.drop_duplicates",
+    "pandas.core.frame.DataFrame.sort_values",
+    "pandas.core.frame.DataFrame.sort_index",
+    "pandas.core.frame.DataFrame.query",
+    "pandas.core.frame.DataFrame.transpose",
+    "pandas.core.frame.DataFrame.swapaxes",
+    "pandas.core.frame.DataFrame.reindex",
+    "pandas.core.frame.DataFrame.reindex_like",
+    "pandas.core.frame.DataFrame.truncate");
```

Comment:
I would just like to mention that I am aware that this list may not be exhaustive. I took it from the RSPEC. Furthermore, some of these functions actually don't have the `inplace` parameter specified in their documentation anymore.

Reply:
As far as I remember, some of these actually come from [PDEP-8](https://github.com/pandas-dev/pandas/pull/51466/files), though I see `eval` is missing and some of the methods from this list are actually not referenced in the PDEP. Which methods do you see that doesn't have the parameter anymore? I think it's fine to have them included if they had the parameter in a previous version of Pandas, as some people might still be using an older version anyway.

That said, I think it's fair to start with this list and re-evaluate if need be. A more radical approach could've been to flag every call to a method on a `pandas.DataFrame` object that uses `inplace=True` due to how misleading the parameter is.

Reply:
As I re-read the PDEP and the method list, I actually think we probably should stick with "group 4" methods that can never operate in place despite the parameter. If the parameter eventually is deprecated, we can then expand to more methods.

The methods you highlight that don't have the `inplace` parameter, actually never had it and have a `copy` parameter instead. I'm not sure why they made it to the list, but I guess it would make sense to remove them to avoid confusion.

Could you update the list accordingly here and in RSPEC? That means the methods that should be flagged are:
`drop`
`dropna`
`drop_duplicates`
`sort_values`
`sort_index`
`eval`
`query`

Sorry for the confusion!

Reply:
I addressed this in d221ad3af4eb334c1cb4ed62de7f937f27cbf717.

Reply:
I also made sure that the error was raised on the argument, rather than the entire qualified expression.